sentinel-adapter/sentinel-apache-dubbo-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/dubbo/SentinelDubboConsumerFilter.java

```diff
@@ -52,33 +55,87 @@ public SentinelDubboConsumerFilter() {
+        }
+
+    }
+
+    private Result syncInvoke(Invoker<?> invoker, Invocation invocation) {
         Entry interfaceEntry = null;
         Entry methodEntry = null;
-        RpcContext rpcContext = RpcContext.getContext();
+        String methodResourceName = getMethodName(invoker, invocation);
+        String interfaceResourceName = getInterfaceName(invoker);
         try {
-            String methodResourceName = DubboUtils.getResourceName(invoker, invocation, DubboConfig.getDubboConsumerPrefix());
-            String interfaceResourceName = DubboConfig.getDubboInterfaceGroupAndVersionEnabled() ? invoker.getUrl().getColonSeparatedKey()
-                    : invoker.getInterface().getName();
-            InvokeMode invokeMode = RpcUtils.getInvokeMode(invoker.getUrl(), invocation);
-
-            if (InvokeMode.SYNC == invokeMode) {
-                interfaceEntry = SphU.entry(interfaceResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.OUT);
-                rpcContext.set(DubboUtils.DUBBO_INTERFACE_ENTRY_KEY, interfaceEntry);
-                methodEntry = SphU.entry(methodResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.OUT, invocation.getArguments());
-            } else {
-                // should generate the AsyncEntry when the invoke model in future or async
-                interfaceEntry = SphU.asyncEntry(interfaceResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.OUT);
-                rpcContext.set(DubboUtils.DUBBO_INTERFACE_ENTRY_KEY, interfaceEntry);
-                methodEntry = SphU.asyncEntry(methodResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.OUT, 1, invocation.getArguments());
+            interfaceEntry = SphU.entry(interfaceResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.OUT);
+            methodEntry = SphU.entry(methodResourceName, ResourceTypeConstants.COMMON_RPC, EntryType.OUT, invocation.getArguments());
+            Result result = invoker.invoke(invocation);
+            if (result.hasException()) {
+                Tracer.traceEntry(result.getException(), interfaceEntry);
+                Tracer.traceEntry(result.getException(), methodEntry);
+            }
+            return result;
+        } catch (BlockException e) {
+            return DubboFallbackRegistry.getConsumerFallback().handle(invoker, invocation, e);
+        } catch (RpcException e) {
+            Tracer.traceEntry(e, interfaceEntry);
+            Tracer.traceEntry(e, methodEntry);
+            throw e;
+        } finally {
+            if (methodEntry != null) {
+                methodEntry.exit();
+            }
+            if (interfaceEntry != null) {
+                interfaceEntry.exit();
             }
-            rpcContext.set(DubboUtils.DUBBO_METHOD_ENTRY_KEY, methodEntry);
-            return invoker.invoke(invocation);
+        }
+    }
+
+
+    private Result asyncInvoke(Invoker<?> invoker, Invocation invocation) {
+        Stack<Entry> entryStack = new Stack<>();
```

Comment:
How about using `Deque` instead to avoid bad performance? `Stack` is derived from `Vector`, whose operations are wrapped with `synchronized`.

Reply:
Oh why not using `final` here? Any performance related reason?

Reply:
final can improve the performance highly?

Reply:
> final can improve the performance highly?

Not really.
Talk to performance in time efficiency i should say there isn't obvious difference. It's a suggestion aiming:

* No extra data structure was introduced.
* Look simpler
* Implicit reference was introduced for `Stack` or `Dequeue` compared to `final`. They are similar in this view.